File: CashuWallet/Core/NIP44.swift
import Foundation
import CryptoKit
import P256K

enum NIP44 {
    enum Error: Swift.Error {
        case invalidPubkey
        case invalidVersion
        case invalidPayload
        case macMismatch
        case lengthMismatch
        case plaintextTooLong
    }

    /// Encrypt `plaintext` with NIP-44 v2. Returns base64 payload.
    static func encrypt(
        plaintext: String,
        senderPrivateKey: Data,
        recipientPubkeyHex: String
    ) throws -> String {
        let conversationKey = try conversationKey(privateKey: senderPrivateKey, pubkeyHex: recipientPubkeyHex)
        var nonce = Data(count: 32)
        let status = nonce.withUnsafeMutableBytes { SecRandomCopyBytes(kSecRandomDefault, 32, $0.baseAddress!) }
        guard status == errSecSuccess else { throw Error.invalidPayload }
        return try encrypt(plaintext: plaintext, conversationKey: conversationKey, nonce: nonce)
    }

    /// Decrypt a NIP-44 v2 base64 payload.
    static func decrypt(
        payload: String,
        recipientPrivateKey: Data,
        senderPubkeyHex: String
    ) throws -> String {
        let conversationKey = try conversationKey(privateKey: recipientPrivateKey, pubkeyHex: senderPubkeyHex)
        return try decrypt(payload: payload, conversationKey: conversationKey)
    }

    // MARK: - Internal

    static func conversationKey(privateKey: Data, pubkeyHex: String) throws -> Data {
        guard let pubkeyBytes = Data(hex: pubkeyHex), pubkeyBytes.count == 32 else {
            throw Error.invalidPubkey
        }
        // x-only -> uncompressed via 0x02 prefix (parity flag — either works for ECDH X).
        var serialized = Data([0x02])
        serialized.append(pubkeyBytes)
        let pub = try P256K.KeyAgreement.PublicKey(dataRepresentation: serialized, format: .compressed)
        let priv = try P256K.KeyAgreement.PrivateKey(dataRepresentation: privateKey)
        let shared = try priv.sharedSecretFromKeyAgreement(with: pub, format: .compressed)
        // libsecp256k1 returns 33 bytes (1 version byte + 32 X). Strip the version.
        let ssBytes = shared.withUnsafeBytes { Data($0) }
        let x = ssBytes.suffix(32)
        let prk = HKDF<CryptoKit.SHA256>.extract(
            inputKeyMaterial: SymmetricKey(data: x),
            salt: Data("nip44-v2".utf8)
        )
        return prk.withUnsafeBytes { Data($0) }
    }

    static func encrypt(plaintext: String, conversationKey: Data, nonce: Data) throws -> String {
        let plaintextBytes = Data(plaintext.utf8)
        guard plaintextBytes.count >= 1, plaintextBytes.count <= 65_535 else {
            throw Error.plaintextTooLong
        }
        let (chachaKey, chachaNonce, hmacKey) = try derive(conversationKey: conversationKey, nonce: nonce)
        let padded = pad(plaintext: plaintextBytes)
        let ciphertext = try chacha20(key: chachaKey, nonce: chachaNonce, data: padded)
        let mac = hmacAAD(key: hmacKey, aad: nonce, message: ciphertext)
        var payload = Data([0x02])
        payload.append(nonce)
        payload.append(ciphertext)
        payload.append(mac)
        return payload.base64EncodedString()
    }

    static func decrypt(payload: String, conversationKey: Data) throws -> String {
        guard let raw = Data(base64Encoded: payload), raw.count >= 1 + 32 + 32 + 32 else {
            throw Error.invalidPayload
        }
        guard raw[0] == 0x02 else { throw Error.invalidVersion }
        let nonce = raw.subdata(in: 1..<33)
        let mac = raw.suffix(32)
        let ciphertext = raw.subdata(in: 33..<(raw.count - 32))
        let (chachaKey, chachaNonce, hmacKey) = try derive(conversationKey: conversationKey, nonce: nonce)
        let expected = hmacAAD(key: hmacKey, aad: nonce, message: ciphertext)
        guard constantTimeEquals(expected, mac) else { throw Error.macMismatch }
        let padded = try chacha20(key: chachaKey, nonce: chachaNonce, data: ciphertext)
        return try unpad(padded: padded)
    }

    // MARK: - HKDF expand → 76 bytes → split

    private static func derive(conversationKey: Data, nonce: Data) throws -> (chachaKey: Data, chachaNonce: Data, hmacKey: Data) {
        let prk = SymmetricKey(data: conversationKey)
        let derived = HKDF<CryptoKit.SHA256>.expand(pseudoRandomKey: prk, info: nonce, outputByteCount: 76)
        let d = derived.withUnsafeBytes { Data($0) }
        return (d.subdata(in: 0..<32), d.subdata(in: 32..<44), d.subdata(in: 44..<76))
    }

    // MARK: - Padding (NIP-44 v2 spec)

    static func pad(plaintext: Data) -> Data {
        let length = plaintext.count
        let paddedLen = paddedLength(for: length)
        var out = Data(capacity: 2 + paddedLen)
        out.append(UInt8((length >> 8) & 0xFF))
        out.append(UInt8(length & 0xFF))
        out.append(plaintext)
        if paddedLen > length {
            out.append(Data(repeating: 0, count: paddedLen - length))
        }
        return out
    }

    static func unpad(padded: Data) throws -> String {
        guard padded.count >= 2 else { throw Error.invalidPayload }
        let len = (Int(padded[0]) << 8) | Int(padded[1])
        guard len >= 1, len <= padded.count - 2 else { throw Error.lengthMismatch }
        let expectedPaddedLen = paddedLength(for: len)
        guard padded.count == 2 + expectedPaddedLen else { throw Error.lengthMismatch }
        let textBytes = padded.subdata(in: 2..<(2 + len))
        guard let s = String(data: textBytes, encoding: .utf8) else { throw Error.invalidPayload }
        return s
    }

    static func paddedLength(for length: Int) -> Int {
        guard length > 32 else { return 32 }
        let nextPower = 1 << (Int(log2(Double(length - 1))) + 1)
        let chunk: Int = nextPower <= 256 ? 32 : nextPower / 8
        return chunk * (((length - 1) / chunk) + 1)
    }

    // MARK: - Crypto primitives

    private static func chacha20(key: Data, nonce: Data, data: Data) throws -> Data {
        // ChaCha20 stream cipher, IETF variant (96-bit nonce, 32-bit counter starting at 0).
        // NIP-44 uses the 12-byte nonce directly with counter=0.
        return try ChaCha20.process(key: key, nonce: nonce, data: data)
    }

    private static func hmacAAD(key: Data, aad: Data, message: Data) -> Data {
        var data = Data()
        data.append(aad)
        data.append(message)
        let mac = HMAC<CryptoKit.SHA256>.authenticationCode(for: data, using: SymmetricKey(data: key))
        return mac.withUnsafeBytes { Data($0) }
    }

    private static func constantTimeEquals(_ a: Data, _ b: Data) -> Bool {
        guard a.count == b.count else { return false }
        var diff: UInt8 = 0
        for i in 0..<a.count { diff |= a[i] ^ b[i] }
        return diff == 0
    }
}

// MARK: - ChaCha20 (IETF RFC 8439) — pure-Swift implementation

enum ChaCha20 {
    static func process(key: Data, nonce: Data, data: Data) throws -> Data {
        precondition(key.count == 32, "ChaCha20 key must be 32 bytes")
        precondition(nonce.count == 12, "ChaCha20 nonce must be 12 bytes")
        var out = Data(count: data.count)
        var counter: UInt32 = 0
        var offset = 0
        while offset < data.count {
            let block = chacha20Block(key: key, counter: counter, nonce: nonce)
            let remaining = data.count - offset
            let take = min(64, remaining)
            for i in 0..<take {
                out[offset + i] = data[offset + i] ^ block[i]
            }
            offset += take
            counter &+= 1
        }
        return out
    }

    private static func chacha20Block(key: Data, counter: UInt32, nonce: Data) -> [UInt8] {
        var state = [UInt32](repeating: 0, count: 16)
        state[0] = 0x61707865
        state[1] = 0x3320646E
        state[2] = 0x79622D32
        state[3] = 0x6B206574
        for i in 0..<8 {
            state[4 + i] = loadLE32(key, offset: i * 4)
        }
        state[12] = counter
        state[13] = loadLE32(nonce, offset: 0)
        state[14] = loadLE32(nonce, offset: 4)
        state[15] = loadLE32(nonce, offset: 8)

        var working = state
        for _ in 0..<10 {
            quarterRound(&working, 0, 4, 8, 12)
            quarterRound(&working, 1, 5, 9, 13)
            quarterRound(&working, 2, 6, 10, 14)
            quarterRound(&working, 3, 7, 11, 15)
            quarterRound(&working, 0, 5, 10, 15)
            quarterRound(&working, 1, 6, 11, 12)
            quarterRound(&working, 2, 7, 8, 13)
            quarterRound(&working, 3, 4, 9, 14)
        }
        for i in 0..<16 { working[i] &+= state[i] }

        var out = [UInt8](repeating: 0, count: 64)
        for i in 0..<16 {
            let w = working[i]
            out[i * 4] = UInt8(w & 0xFF)
            out[i * 4 + 1] = UInt8((w >> 8) & 0xFF)
            out[i * 4 + 2] = UInt8((w >> 16) & 0xFF)
            out[i * 4 + 3] = UInt8((w >> 24) & 0xFF)
        }
        return out
    }

    private static func loadLE32(_ data: Data, offset: Int) -> UInt32 {
        return UInt32(data[data.startIndex + offset]) |
               (UInt32(data[data.startIndex + offset + 1]) << 8) |
               (UInt32(data[data.startIndex + offset + 2]) << 16) |
               (UInt32(data[data.startIndex + offset + 3]) << 24)
    }

    private static func quarterRound(_ s: inout [UInt32], _ a: Int, _ b: Int, _ c: Int, _ d: Int) {
        s[a] = s[a] &+ s[b]; s[d] ^= s[a]; s[d] = rotl(s[d], 16)
        s[c] = s[c] &+ s[d]; s[b] ^= s[c]; s[b] = rotl(s[b], 12)
        s[a] = s[a] &+ s[b]; s[d] ^= s[a]; s[d] = rotl(s[d], 8)
        s[c] = s[c] &+ s[d]; s[b] ^= s[c]; s[b] = rotl(s[b], 7)
    }

    private static func rotl(_ v: UInt32, _ n: UInt32) -> UInt32 {
        (v << n) | (v >> (32 - n))
    }
}

File: android/app/src/main/java/com/cashu/me/Core/NIP44.kt
package com.cashu.me.Core

import java.math.BigInteger
import java.security.MessageDigest
import java.security.SecureRandom
import java.util.Base64
import javax.crypto.Mac
import javax.crypto.spec.SecretKeySpec
import org.bouncycastle.asn1.sec.SECNamedCurves

internal object NIP44 {
    private val secureRandom = SecureRandom()
    private val params = SECNamedCurves.getByName("secp256k1")
    private val curve = params.curve
    private val order = params.n

    fun encrypt(
        plaintext: String,
        senderPrivateKey: ByteArray,
        recipientPubkeyHex: String,
    ): String {
        val conversationKey = conversationKey(senderPrivateKey, recipientPubkeyHex)
        val nonce = ByteArray(32).also(secureRandom::nextBytes)
        return encrypt(plaintext, conversationKey, nonce)
    }

    fun decrypt(
        payload: String,
        recipientPrivateKey: ByteArray,
        senderPubkeyHex: String,
    ): String {
        val conversationKey = conversationKey(recipientPrivateKey, senderPubkeyHex)
        return decrypt(payload, conversationKey)
    }

    internal fun conversationKey(privateKey: ByteArray, pubkeyHex: String): ByteArray {
        val pubkeyBytes = hexToBytes(pubkeyHex)
        require(pubkeyBytes.size == 32) { "Invalid Nostr public key." }
        val secret = privateKey.toPositiveBigInteger()
        require(privateKey.size == 32 && secret > BigInteger.ZERO && secret < order) { "Invalid Nostr private key." }
        val publicPoint = curve.decodePoint(byteArrayOf(0x02) + pubkeyBytes)
        val sharedPoint = publicPoint.multiply(secret).normalize()
        val sharedX = sharedPoint.affineXCoord.encoded.toFixedSize(32)
        return hmacSha256("nip44-v2".toByteArray(Charsets.UTF_8), sharedX)
    }

    internal fun encrypt(plaintext: String, conversationKey: ByteArray, nonce: ByteArray): String {
        val plaintextBytes = plaintext.toByteArray(Charsets.UTF_8)
        require(plaintextBytes.size in 1..65_535) { "NIP-44 plaintext size is invalid." }
        val (chachaKey, chachaNonce, hmacKey) = derive(conversationKey, nonce)
        val padded = pad(plaintextBytes)
        val ciphertext = ChaCha20.process(chachaKey, chachaNonce, padded)
        val mac = hmacSha256(hmacKey, nonce + ciphertext)
        val payload = byteArrayOf(0x02) + nonce + ciphertext + mac
        return Base64.getEncoder().encodeToString(payload)
    }

    internal fun decrypt(payload: String, conversationKey: ByteArray): String {
        val raw = Base64.getDecoder().decode(payload)
        require(raw.size >= 65) { "Invalid NIP-44 payload." }
        require(raw[0] == 0x02.toByte()) { "Invalid NIP-44 version." }
        val nonce = raw.copyOfRange(1, 33)
        val ciphertext = raw.copyOfRange(33, raw.size - 32)
        val mac = raw.copyOfRange(raw.size - 32, raw.size)
        val (chachaKey, chachaNonce, hmacKey) = derive(conversationKey, nonce)
        val expected = hmacSha256(hmacKey, nonce + ciphertext)
        require(constantTimeEquals(expected, mac)) { "NIP-44 MAC mismatch." }
        val padded = ChaCha20.process(chachaKey, chachaNonce, ciphertext)
        return unpad(padded)
    }

    internal fun pad(plaintext: ByteArray): ByteArray {
        val paddedLength = paddedLength(plaintext.size)
        return ByteArray(2 + paddedLength).also { out ->
            out[0] = ((plaintext.size ushr 8) and 0xff).toByte()
            out[1] = (plaintext.size and 0xff).toByte()
            plaintext.copyInto(out, destinationOffset = 2)
        }
    }

    internal fun unpad(padded: ByteArray): String {
        require(padded.size >= 2) { "Invalid NIP-44 padding." }
        val length = ((padded[0].toInt() and 0xff) shl 8) or (padded[1].toInt() and 0xff)
        require(length in 1..(padded.size - 2)) { "Invalid NIP-44 length." }
        require(padded.size == 2 + paddedLength(length)) { "Invalid NIP-44 padded length." }
        return padded.copyOfRange(2, 2 + length).toString(Charsets.UTF_8)
    }

    internal fun paddedLength(length: Int): Int {
        require(length > 0) { "NIP-44 plaintext must not be empty." }
        if (length <= 32) return 32
        val nextPower = Integer.highestOneBit(length - 1) shl 1
        val chunk = if (nextPower <= 256) 32 else nextPower / 8
        return chunk * (((length - 1) / chunk) + 1)
    }

    private data class DerivedKeys(
        val chachaKey: ByteArray,
        val chachaNonce: ByteArray,
        val hmacKey: ByteArray,
    )

    private fun derive(conversationKey: ByteArray, nonce: ByteArray): DerivedKeys {
        require(conversationKey.size == 32) { "NIP-44 conversation key must be 32 bytes." }
        require(nonce.size == 32) { "NIP-44 nonce must be 32 bytes." }
        val derived = hkdfExpand(conversationKey, nonce, 76)
        return DerivedKeys(
            chachaKey = derived.copyOfRange(0, 32),
            chachaNonce = derived.copyOfRange(32, 44),
            hmacKey = derived.copyOfRange(44, 76),
        )
    }

    private fun hkdfExpand(prk: ByteArray, info: ByteArray, length: Int): ByteArray {
        val output = ByteArray(length)
        var previous = ByteArray(0)
        var offset = 0
        var counter = 1
        while (offset < length) {
            previous = hmacSha256(prk, previous + info + counter.toByte())
            val take = minOf(previous.size, length - offset)
            previous.copyInto(output, destinationOffset = offset, endIndex = take)
            offset += take
            counter += 1
        }
        return output
    }

    private fun hmacSha256(key: ByteArray, data: ByteArray): ByteArray {
        val mac = Mac.getInstance("HmacSHA256")
        mac.init(SecretKeySpec(key, "HmacSHA256"))
        return mac.doFinal(data)
    }

    private fun constantTimeEquals(a: ByteArray, b: ByteArray): Boolean {
        if (a.size != b.size) return false
        var diff = 0
        for (index in a.indices) {
            diff = diff or (a[index].toInt() xor b[index].toInt())
        }
        return diff == 0
    }

    internal fun hexToBytes(hex: String): ByteArray {
        val clean = hex.trim().removePrefix("0x")
        require(clean.length % 2 == 0) { "Hex string must have an even length." }
        return clean.chunked(2).map { it.toInt(16).toByte() }.toByteArray()
    }

    private fun ByteArray.toPositiveBigInteger(): BigInteger = BigInteger(1, this)

    private fun ByteArray.toFixedSize(size: Int): ByteArray {
        val positive = if (isNotEmpty() && first() == 0.toByte()) drop(1).toByteArray() else this
        require(positive.size <= size) { "Value does not fit in $size bytes." }
        return ByteArray(size - positive.size) + positive
    }
}

private object ChaCha20 {
    fun process(key: ByteArray, nonce: ByteArray, data: ByteArray): ByteArray {
        require(key.size == 32) { "ChaCha20 key must be 32 bytes." }
        require(nonce.size == 12) { "ChaCha20 nonce must be 12 bytes." }
        val out = ByteArray(data.size)
        var counter = 0
        var offset = 0
        while (offset < data.size) {
            val block = block(key, counter, nonce)
            val take = minOf(64, data.size - offset)
            for (index in 0 until take) {
                out[offset + index] = (data[offset + index].toInt() xor block[index].toInt()).toByte()
            }
            offset += take
            counter += 1
        }
        return out
    }

    private fun block(key: ByteArray, counter: Int, nonce: ByteArray): ByteArray {
        val state = IntArray(16)
        state[0] = 0x61707865
        state[1] = 0x3320646e
        state[2] = 0x79622d32
        state[3] = 0x6b206574
        for (index in 0 until 8) {
            state[4 + index] = loadLE32(key, index * 4)
        }
        state[12] = counter
        state[13] = loadLE32(nonce, 0)
        state[14] = loadLE32(nonce, 4)
        state[15] = loadLE32(nonce, 8)
        val working = state.copyOf()
        repeat(10) {
            quarterRound(working, 0, 4, 8, 12)
            quarterRound(working, 1, 5, 9, 13)
            quarterRound(working, 2, 6, 10, 14)
            quarterRound(working, 3, 7, 11, 15)
            quarterRound(working, 0, 5, 10, 15)
            quarterRound(working, 1, 6, 11, 12)
            quarterRound(working, 2, 7, 8, 13)
            quarterRound(working, 3, 4, 9, 14)
        }
        for (index in 0 until 16) {
            working[index] += state[index]
        }
        val out = ByteArray(64)
        for (index in 0 until 16) {
            val word = working[index]
            out[index * 4] = (word and 0xff).toByte()
            out[index * 4 + 1] = ((word ushr 8) and 0xff).toByte()
            out[index * 4 + 2] = ((word ushr 16) and 0xff).toByte()
            out[index * 4 + 3] = ((word ushr 24) and 0xff).toByte()
        }
        return out
    }

    private fun quarterRound(state: IntArray, a: Int, b: Int, c: Int, d: Int) {
        state[a] += state[b]
        state[d] = Integer.rotateLeft(state[d] xor state[a], 16)
        state[c] += state[d]
        state[b] = Integer.rotateLeft(state[b] xor state[c], 12)
        state[a] += state[b]
        state[d] = Integer.rotateLeft(state[d] xor state[a], 8)
        state[c] += state[d]
        state[b] = Integer.rotateLeft(state[b] xor state[c], 7)
    }

    private fun loadLE32(data: ByteArray, offset: Int): Int =
        (data[offset].toInt() and 0xff) or
            ((data[offset + 1].toInt() and 0xff) shl 8) or
            ((data[offset + 2].toInt() and 0xff) shl 16) or
            ((data[offset + 3].toInt() and 0xff) shl 24)
}
